packages/devtools_app/lib/src/memory/memory_instance_tree_view.dart

```diff
@@ -0,0 +1,166 @@
+    //              controller. Have other ValueListenables on controller to
+    //              listen to, so we don't need the setState calls.
+    // Update the chart when the memorySource changes.
+    addAutoDisposeListener(controller.selectedSnapshotNotifier, () {
+      setState(() {
+        controller.computeAllLibraries(rebuild: true);
+      });
+    });
+  }
+
+  void setupColumns() {
+    columns.addAll([
+      treeColumn,
+      _FieldNameColumn(),
+      _FieldValueColumn(),
+    ]);
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    controller.instanceFieldsTreeTable = TreeTable<FieldReference>(
+      dataRoots: controller.instanceRoot,
+      columns: columns,
+      treeColumn: treeColumn,
+      keyFactory: (typeRef) => PageStorageKey<String>(typeRef.name),
+      sortColumn: columns[0],
+      sortDirection: SortDirection.ascending,
+    );
+
+    return controller.instanceFieldsTreeTable;
+  }
+}
+
+class _FieldTypeColumn extends TreeColumnData<FieldReference> {
+  _FieldTypeColumn() : super('Type');
+
+  @override
+  dynamic getValue(FieldReference dataObject) =>
+      dataObject.isEmptyReference || dataObject.isSentinelReference
+          ? ''
+          : dataObject.type;
+
+  @override
+  String getDisplayValue(FieldReference dataObject) =>
+      '${getValue(dataObject)}';
+
+  @override
+  bool get supportsSorting => true;
+
+  @override
+  int compare(FieldReference a, FieldReference b) {
+    final Comparable valueA = getValue(a);
+    final Comparable valueB = getValue(b);
+    return valueA.compareTo(valueB);
+  }
+
+  @override
+  double get fixedWidthPx => 250.0;
+}
+
+class _FieldNameColumn extends ColumnData<FieldReference> {
+  _FieldNameColumn() : super('Name');
+
+  @override
+  dynamic getValue(FieldReference dataObject) =>
+      dataObject.isEmptyReference ? '' : dataObject.name;
+
+  @override
+  String getDisplayValue(FieldReference dataObject) =>
+      '${getValue(dataObject)}';
+
+  @override
+  bool get supportsSorting => true;
+
+  @override
+  int compare(FieldReference a, FieldReference b) {
+    final Comparable valueA = getValue(a);
+    final Comparable valueB = getValue(b);
+    return valueA.compareTo(valueB);
+  }
+
+  @override
+  double get fixedWidthPx => 150.0;
+}
+
+class _FieldValueColumn extends ColumnData<FieldReference> {
+  _FieldValueColumn() : super('Value');
+
+  @override
+  dynamic getValue(FieldReference dataObject) =>
+      dataObject.isEmptyReference || dataObject.isSentinelReference
+          ? ''
+          : dataObject.value;
+
+  @override
+  String getDisplayValue(FieldReference dataObject) {
+    if (dataObject is ObjectFieldReference && !dataObject.isNull) {
+      // Real object that isn't Null value is empty string.
+      return '';
+    }
+
+    var value = getValue(dataObject);
+    if (value is String && value.length > 30) {
+      value = '${value.substring(0, 13)}…${value.substring(value.length - 17)}';
```

Comment:
why this is needed?
It would be nice to use standard Flutter string truncation with `...`
If truncating in the middle gives a much better result, I would prefer for us to at some point implement it robustly with text measurement so we actually truncate the correct # of characters rather than guessing that we can only show 30 characters.

Reply:
// TODO(terry): Can we use standard Flutter string truncation?
    //              Long variable names the beginning and end of the name is
    //              most significant the middle part of name less so.  However,
    //              we need a nice hover/tooltip to show the entire name.
    //              If middle, '...' then we need a more robust computation w/
    //              text measurement so we accurately truncate the correct # of
    //              characters instead guessing that we only show 30 chars.